### DISCUSSION TITLE
Support SwiftUI `Table` in SkipUI

### DISCUSSION THREAD

**Discussion by wsdev-co** [skiptools/skip#145] [General]
Dear All,

We're currently developing ERP/BOS platform (similar to Salesforce, SAP, etc.), which is data-dense industry, and softwares are required to have tables in order to show business-related information. Hence, I wanted to know if there's any chance to support [Table](https://developer.apple.com/documentation/swiftui/table) view of SwiftUI to Kotlin, Android?
We are new company, but we are ready to buy `Professional` license of Skip. Currently, we have released 3 iOS, 3 iPadOS, 3 macOS versions to AppStore with 4 localizations, and number of softwares will reach to 7 at the end of September, but we must start transpiling them to Kotlin from now. If there's schedule to support point, we would appreciate to hear.

Regards

**Comment by aabewhite** [comment]
We generally prioritize features based on requests like this one. So while we didn't have immediate plans to implement Table, we can probably do so in the several weeks if it's critical to your needs (unfortunately it's summer vacation time, so the next 2 weeks are only for critical maintenance and support). 

It would be useful to know what features of Table you use?

Also: have you already evaluated Skip? If haven't already done so, I would encourage you to do that now, and just assume that Table can be implemented (I don't see any reason it couldn't be). You could for example temporarily represent the data with a simple List during your evaluation. Skip is great for new apps, but porting a large existing app to Android is not easy, even with the help of Skip. Our Getting Started documentation about migrating existing apps talks about this topic. I want to be sure that you have some experience with Skip before we rush into custom work and you rush into buying a license.

**Reply by guitaripod** [reply]
really appreciate this reply 👍 insightful

**Reply by lordabrahams** [reply]
Thanks a lot to clear reply. 

Let me describe my context,

I have 3 years of advanced experience on Python & Swift, 2 years on Kotlin. My main focus is in system/back-end based development and I developed my own network-protocol & framework (SDK, Driver, etc), but also built 8 Swift/Kotlin-based software apps. 
I'm team-leader of my software company and currently, I fired many team-members due to "laziness"-reasons, and now, until September 2024, I have to complete 6 software for iOS, iPadOS, macOS, Android, Windows platforms, which in total of 30 apps, by myself. So far 3 software for iOS, iPadOS, macOS completed, and released to the AppStore. In next 10 days will complete all Swift-based software, then I start Kotlin-based software apps. 
Hence, it would be great if our `Table` is available even after 2 weeks, until that, as you said, I will mock the `Table` with `List` and finish the detail views. 

[Link](https://apps.apple.com/us/app/dux-pos/id6479682508) to released apps to AppStore so far.

I attached screenshots of `Table` usage in our apps down below. Mainly, we use common features of `Table` such as `.width(min:ideal:max:)`, `init(title:content:)`. There's some advanced features for "customization", "reordering columns", "sorting the rows", which we never intended to use. We've done these functions separately.
 
![IMG_8405](https://github.com/skiptools/skip/assets/87901333/04068086-b43c-4a4a-bc6b-2316b5809ae0)
![Screenshot 2024-06-17 at 7 41 53 PM](https://github.com/skiptools/skip/assets/87901333/b59e47ff-774e-4519-9e62-0b6279b69224)

**Reply by lordabrahams** [reply]
Attachments for previous reply:
![Screenshot 2024-06-18 at 1 55 53 AM](https://github.com/skiptools/skip/assets/87901333/7f3ad449-ee4c-4b0f-a40b-90a045c56fad)
![Screenshot 2024-06-18 at 1 56 49 AM](https://github.com/skiptools/skip/assets/87901333/46803e42-3610-46ab-a03d-3bb17446bbfb)

**Reply by lordabrahams** [reply]
@aabewhite 

Hello Abe, 


I hope summer vacation is going well, I wanted to know if you can let me know the approximate range of date of SwiftUI `Table` support in `SkipUI` framework?

Reason why I am in rush is, I have upcoming meetings with investors, which I have to clarify the future development plan. 


Regards

**Comment by aabewhite** [comment]
I’m traveling for another couple of days. After that I should be able to prototype Table to get a sense for how long it will take to implement. So I should be able to provide an estimate later this week.

**Reply by wsdev-co** [reply]
Thanks a lot, that's great. Have a nice travel.

**Reply by aabewhite** [reply]
Hi - I'm back and should be able to explore Table a bit tomorrow. Are you targeting this for Android tablets, or for phones? I ask because Table's documentation indicates that it behaves very differently on horizontally-constrained devices. 

There are also many other aspects of tablet UI that Skip doesn't yet support, like split views and disclosure groups. If you're targeting Android phones, you can work around these issues by providing a simplified Android UI. But if you're targeting Android tablets and want an iPad-like experience, I'm afraid that Skip currently lacks a lot of the UI elements you'd need.

**Reply by wsdev-co** [reply]
Hello Abe,


That sounds great. I am targeting both phones and tablets. You're right, it behaves as `List` view in small devices, as `Table` in large devices.

Split views wouldn't be required, I would mock the split views with `HStack` (`Column` in context of Kotlin) and configure the software to fixed horizontal rotation (e.g. Games, etc.). This is totally normal for us.

And there was another subject which is little bit context-intensive and weeks ago I wanted to contact with you since I saw your YouTube videos, but I accidentally contacted with Marc. Hence, can you give me your email, I would email the obstacle?
 
 
Regards

**Reply by aabewhite** [reply]
abe@skip.tools

**Reply by aabewhite** [reply]
I prototyped to the point that I know that we can get some basic Table functionality. The following is what I got working for the prototype: 

```
private struct TableData: Identifiable {
    let id = UUID()
    let name: String
    let value: Int
}

private struct TablePlayground: View {
    @State var data = Self.initialData

    var body: some View {
        Table(data) {
            TableColumn("Name", value: \.name)
            TableColumn("Value") { data in
                Text("\(data.value)")
            }
        }
    }

    static var initialData: [TableData] {
        return (1...20).map { TableData(name: "Item \($0)", value: $0) }
    }
}
```

I can't be sure of which other features we can support exactly without implementing those features, which will take a significant amount of effort. 

What is your timeline?

**Reply by wsdev-co** [reply]
This is great, the basic features we need. If you've noticed, there's one `Table` initializer which goes `init(..., selection: Set<Hashable>, ...)`. This initializer is used all the time. Let me share our one of the most simple view snippet code:

```
struct CustomerView: View {
    
    @StateObject var mvvm: AccountMVVM
    
    var body: some View {
        ConsumeView(mvvm: mvvm) { selection in
            Table(mvvm.content, selection: selection) {
                TableColumn("Full name") { each in 
                    Text(verbatim: each.full_name)
                            .pagination(of: mvvm, element: each) 
                 }.width(ideal: 240)
                TableColumn("Credential", value: \.credential).width(ideal: 200)
                
                if is_mac {
                    TableColumn("Last modified", value: \.revision.description).width(ideal: 240)
                    TableColumn("Created at", value: \.creation.description).width(ideal: 240)
                    TableColumn("Deleted", value: \.deletion.description).width(ideal: 100)
                }
            }
        } create: {
            AccountDetailView(intention: .create, mvvm: AccountDetailMVVM())
        } detail: { value in
            AccountDetailView(intention: .update, mvvm: AccountDetailMVVM(focus: value))
        }
    }
}
``` 

Code context:
- The `ConsumeView(mvvm: BaseMVVM)` is the representer of `BaseMVVM` which our `cmos-sdk-swift` provides. And base MVVM is responsible to calls to back-end through `cmos-driver-swift` and it automatically changes the flow to retry views in case of errors, automatically paginates, reacts to the real-time change events and more.
- Here purpose of using `selection` is navigation. When user selected one or more rows, our `ConsumeView` will navigate to only first row.
- Here table modifier `.width(ideal: Double)` used, which configures the preferred table width on appear, then user can further adjust. This is not critical if you cannot implement it in Kotlin, but would be wonderful to have. 
- The `is_mac` is simple flag indicates the current running OS.
- We put key informational column to first, in this case `Full name` which will be the title of `List` items in iPhone/small devices.

Outcome:
We need one critical feature
- `selection: Set<Hashable>` of table [initializer](https://developer.apple.com/documentation/swiftui/table/init(_:selection:columns:)-758ug).
and one preferred feature
- `.width(ideal: Double)` modifier.


Timeline is complete 30 software mentioned [above](https://github.com/orgs/skiptools/discussions/145#discussioncomment-9796290) until 1-september. If you can implement those table features in few weeks, then it would be wonderful. I would discuss the changing the our deadline around Mid-September.

**Reply by aabewhite** [reply]
Basic Table support has been merged, including selection and some .width support. Note that .width(ideal:) isn't really going to do anything on Android, but .width(_:) and .width(min:, max:) are honored.
